server/platform/services/sharedchannel/channelunshare.go

```
@@ -0,0 +1,129 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sharedchannel
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/platform/services/remotecluster"
+)
+
+// channelUnshareMsg represents a notification to a remote cluster to unshare a channel.
+type channelUnshareMsg struct {
+	ChannelId string `json:"channel_id"`
+	RemoteId  string `json:"remote_id"`
+}
+
+// SendChannelUnshare asynchronously sends an unshare notification to a remote cluster.
```

Although this approach works well, it does so in a synchronous way (we remove a share and the server sends a notification to the remotes). It'd be better to implement it in a reactive way, as in:
- Server A removes the share, nothing is sent to any remote, the sync process simply stops.
- Server B sends a sync payload for the removed share, as it is unaware of the removal.
- Server A receives the payload, and returns a specific error indicating that the share is not active anymore.
- Server B receives the error in the callback of the sync, and decides that it wants to remove the share as a consequence.

This way, it doesn't matter the state of the connection at the time of the action; the system eventually arrives to a consistent state. It avoids as well the implementation of new messages, new handlers, etc, as we only need to return a specific error and process it in the sync callback, simplifying the code as well.

---

In general that is indeed better - it also aligns with the current architecture which is mostly async, non-blocking, event-driven. 
In practice, I don't think there is much difference, the eventual consistency given by the async approach should happen relatively fast given that the sync operations typically occur frequently I'd guess, so the system should get into a good state  relatively fast with both approaches (delayed feedback on Server B UI is the only real user experience difference I think  but I don't know if it matters).

---

My main concern with the sync approach is that if the servers happen to be disconnected (or if the server removing the share is restarted), we have to create queues, store the action intent and repeat it, which adds plenty of complexity on top of the new listeners/handlers.

The usage experience would be similar, I totally agree, the sync approach would even achieve consistency faster if the machines are connected, but I think it's worth going async whenever we can to maintain the pattern and for code simplicity.